src/http/http2_types.h

```diff
@@ -19,12 +20,60 @@ namespace http2
 
   constexpr static size_t max_data_read_size = 1 << 20;
 
+  // Used to keep track of response state between nghttp2 callbacks and to
+  // differentiate unary from streaming responses
+  enum class StreamResponseState
+  {
+    Uninitialised = 0, // No response to send yet
+    Closing, // Unary or last frame in stream
+    Streaming // Sending data frames to client
+  };
+
+  class DataSource
+  {
+    // Utility class to consume data from underlying data vector in chunks from
+    // nghttp2_data_source_read_callback
+  private:
+    std::vector<uint8_t> data;
+
+    std::span<const uint8_t> span;
```

Comment:
Not sure why we have the span here? That looks like the ownership is with data, we can return a fresh span when necessary?

Reply:
For posterity:

This struct is used to store data we're trying to `send` through http2. When http2 calls us back to copy this data, its target buffer has a capped size. So if we're trying to send a 10MB `data`, we may get called multiple times and told to copy in just 4K at a time. From our perspective, this all happens under a single call we made, but http2 insists on chunking it per-frame or something. When that happens, instead of truncating the start of `data` each time (with potential reallocation/realignment costs) to indicate how much we've written, or maintaining a separate `offset` counting how many bytes we've written, we thought it was neater to use a `span` - it always points at a subrange (and in fact suffix) of `data`, but is cheap to trim the start off of as we advance.

Reply:
@jumaffre can we put what @eddyashton wrote in the comment above? :)